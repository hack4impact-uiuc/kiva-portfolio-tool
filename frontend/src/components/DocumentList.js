import React, { Component } from 'react'
import { Table } from 'reactstrap'

import { connect } from 'react-redux'

import DocumentListItem from './DocumentListItem'

import expand from '../media/expand.png'
import collapse from '../media/collapse.png'

import '../styles/documentlist.scss'

const mapStateToProps = state => ({
  language: state.user.language
})

/**
 * Displays all documents for a given document status in its own box
 */
export class DocumentList extends Component {
  constructor(props) {
    super(props)

    this.state = {
      expanded: false,
      /**
       * showLess is the maximum number to be shown before an expand button appears
       */
      showLess: 5
    }
  }

  /**
   * Expands the number of documents shown only if there are more than 5 documents in current
   * document class
   */
  toggleExpand = () => {
    this.setState({ expanded: !this.state.expanded })
  }

  languages = {
    English: {
      statuses: {
        Missing: 'Missing',
        Pending: 'Pending',
        Rejected: 'Rejected',
        Approved: 'Approved'
      },
      documents: ' documents',
      documentName: 'Document name',
      file: 'File'
    },
    Spanish: {
      statuses: {
        Missing: 'Incompleto',
        Pending: 'Pendiente',
        Rejected: 'Rechazado',
        Approved: 'Aprobado'
      },
      documents: ' documentos',
      documentName: 'Nombre de documento',
      file: 'Archivo'
    },
    French: {
      statuses: {
        Missing: 'Manquant',
        Pending: 'En attente',
        Rejected: 'Rejeté',
        Approved: 'Approuvé'
      },
      documents: ' documents',
      documentName: 'Nom du document',
      file: 'Fichier'
    },
    Portuguese: {
      statuses: {
        Missing: 'Falta',
        Pending: 'Pendente ',
        Rejected: 'Rejeitado',
        Approved: 'Aprovado'
      },
      documents: ' documentos',
      documentName: 'Nome do documento',
      file: 'Arquivo'
    }
  }

  render() {
    let text = this.languages[this.props.language]
    if (!text) {
      text = this.languages['English']
    }

    return (
      <Table className="document-table">
        <tbody>
          <tr className="dlist-header">
            <th
              colSpan="3"
              className={
                'background-' +
                (this.props.status ? this.props.status.toLowerCase() : 'null') +
                ' text-white'
              }
            >
              {text.statuses[this.props.status] + text.documents}
            </th>
          </tr>
          <tr>
            <th className="text-centered">{text.documentName}</th>
            <th className="text-centered">{text.file}</th>
            <th />
          </tr>
          {this.props.documents
            ? this.state.expanded
              ? this.props.documents.map((document, index) => (
                  /* using index should differentiate the documents from the siblings */
                  <DocumentListItem
                    key={index}
                    document={document}
                    docClass={document.docClass}
                    fileName={document.fileName}
                    fileId={document.fileId}
                  />
                ))
              : this.props.documents
                  .slice(0, this.state.showLess)
                  .map((document, index) => (
                    <DocumentListItem
                      key={index}
                      document={document}
                      docClass={document.docClass}
                      fileName={document.fileName}
                      fileId={document.fileId}
                    />
                  ))
            : null}
          {this.props.documents ? (
            this.props.documents.length > this.state.showLess ? (
              <tr className="text-centered">
                <td colSpan="3" id="buttonRowData">
                  {this.state.expanded ? (
                    <button
                      className={
                        'expandButton ' +
                        'background-' +
                        (this.props.status ? this.props.status.toLowerCase() : 'null')
                      }
                      onClick={this.toggleExpand}
                    >
                      <img className="expandButtonImage" src={collapse} alt="Expand icon" />
                    </button>
                  ) : (
                    <button
                      className={
                        'expandButton ' +
                        'background-' +
                        (this.props.status ? this.props.status.toLowerCase() : 'null')
                      }
                      onClick={this.toggleExpand}
                    >
                      <img className="expandButtonImage" src={expand} alt="Expand icon" />
                    </button>
                  )}
                </td>
              </tr>
            ) : null
          ) : null}
        </tbody>
      </Table>
    )
  }
}

export default connect(mapStateToProps)(DocumentList)
