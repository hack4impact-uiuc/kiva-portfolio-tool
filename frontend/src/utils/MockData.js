const MockData = {
  missing: [
    {
      status: 'Missing',
      docClass: 'Board of Directors'
    },
    {
      status: 'Missing',
      docClass: 'Auditor Letter'
    }
  ],
  pending: [
    {
      status: 'Pending',
      docClass: 'Income Statement',
      fileName: 'income_statement.pdf',
      uploadDate: '12/4/2018'
    },
    {
      status: 'Pending',
      docClass: 'Balance Sheet',
      fileName: 'balance_sheet.pdf',
      uploadDate: '1/24/2019'
    }
  ],

  rejected: [
    {
      status: 'Rejected',
      docClass: 'Strategic Plan',
      fileName: 'strategic_plan.pdf',
      uploadDate: '12/24/2018'
    },
    {
      status: 'Rejected',
      docClass: 'Annual Plan',
      fileName: 'annual_plan.pdf',
      uploadDate: '1/3/2019'
    }
  ],
  approved: [
    {
      status: 'Approved',
      docClass: 'Financial Projections',
      fileName: 'financial_proj.pdf',
      uploadDate: '12/24/2018'
    },
    {
      status: 'Approved',
      docClass: 'Organizational Chart',
      fileName: 'org_chard.pdf',
      uploadDate: '1/23/2019'
    }
  ]
}

export default MockData
