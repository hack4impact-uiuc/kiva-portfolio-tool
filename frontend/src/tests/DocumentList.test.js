import React from 'react'
import DocumentList from '../components/DocumentList'
import Adapter from 'enzyme-adapter-react-16'
import { shallowToJson } from 'enzyme-to-json'
import { configure, shallow } from 'enzyme'
import MockData from '../utils/MockData'

beforeAll(() => {
  configure({ adapter: new Adapter() })
})

test('DocumentList renders correctly when there are no documents', () => {
  const component = shallow(<DocumentList />)

  let tree = shallowToJson(component)
  expect(tree).toMatchSnapshot()
})

test('DocumentList renders correctly with documents', () => {
  const component = shallow(<DocumentList documents={MockData.missing} />)

  let tree = shallowToJson(component)
  expect(tree).toMatchSnapshot()
})
