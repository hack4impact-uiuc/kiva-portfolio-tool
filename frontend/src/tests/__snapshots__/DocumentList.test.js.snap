// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentList renders correctly when there are no documents 1`] = `
<Table
  className="document-table"
  responsiveTag="div"
  tag="table"
>
  <tbody>
    <tr
      className="dlist-header"
    >
      <th
        className="background-null text-white"
        colSpan="3"
      >
        undefined documents
      </th>
    </tr>
    <tr>
      <th
        className="text-centered"
      >
        Document name
      </th>
      <th
        className="text-centered"
      >
        File
      </th>
      <th />
    </tr>
  </tbody>
</Table>
`;

exports[`DocumentList renders correctly with documents 1`] = `
<Table
  className="document-table"
  responsiveTag="div"
  tag="table"
>
  <tbody>
    <tr
      className="dlist-header"
    >
      <th
        className="background-null text-white"
        colSpan="3"
      >
        undefined documents
      </th>
    </tr>
    <tr>
      <th
        className="text-centered"
      >
        Document name
      </th>
      <th
        className="text-centered"
      >
        File
      </th>
      <th />
    </tr>
    <Connect(withRouter(DocumentListItem))
      docClass="Board of Directors"
      document={
        Object {
          "docClass": "Board of Directors",
          "status": "Missing",
        }
      }
      key="0"
    />
    <Connect(withRouter(DocumentListItem))
      docClass="Auditor Letter"
      document={
        Object {
          "docClass": "Auditor Letter",
          "status": "Missing",
        }
      }
      key="1"
    />
  </tbody>
</Table>
`;
